pylabrobot/pumps/agrowpumps/agrowdosepump.py

```diff
@@ -0,0 +1,185 @@
+      self.modbus = SimulatedModbusClient()
+    else:
+      self.modbus = ModbusClient(method="rtu", port=self.port,
+                                 baudrate=115200, timeout=1, stopbits=1, bytesize=8, parity="E",
+                                 retry_on_empty=True)
+    assert self.modbus is not None, "Modbus connection not established"
+    await self.modbus.connect()  # type: ignore
+    self.start_keep_alive_thread()
+    self.pump_index_to_address = {pump: pump + 100 for pump in range(self.num_channels)}
+
+  async def run_revolutions(self, num_revolutions: List[float],
+                            use_channels: List[int]):
+    """Run the specified channels at the speed selected. If speed is 0, the pump will be halted.
+      Args:
+        num_revolutions: number of revolutions to run pumps.
+        use_channels: pump array channels to run
+
+      Raises:
+        NotImplementedError: Revolution based pumping commands are not available for this array.
+    """
+    raise NotImplementedError(
+      "Revolution based pumping commands are not available for this pump array.")
+
+  async def run_continuously(self, speed: List[float],
+                             use_channels: List[int]):
+    """Run pumps at the specified speeds.
+        Args:
+            speed: rate at which to run pump.
+            use_channels: pump array channels to run
+
+        Raises:
+            ValueError: Pump address out of range
+            ValueError: Pump speed out of range
+    """
+    assert self.modbus is not None, "Modbus connection not established"
+    assert self.pump_index_to_address is not None, "Pump address mapping not established"
+    if any(channel not in range(1, self.num_channels + 1) for channel in use_channels):
+      raise ValueError(f"Pump address out of range for this pump array. \
+        Value should be between 1 and {self.num_channels+1}")
+    for pump_index, pump_speed in zip(use_channels, speed):
+      pump_speed = int(pump_speed)
+      if pump_speed not in range(101):
+        raise ValueError("Pump speed out of range. Value should be between 0 and 100.")
+      self.modbus.write_register(self.pump_index_to_address[pump_index], pump_speed, unit=self.unit)
+
+  async def halt(self):
+    """ Halt the entire pump array. """
+    assert self.modbus is not None, "Modbus connection not established"
+    assert self.pump_index_to_address is not None, "Pump address mapping not established"
+    print("Engaging shutdown routine")
+    for pump in self.pump_index_to_address:
+      address = self.pump_index_to_address[pump]
+      self.modbus.write_register(address, 0, unit=self.unit)
+
+  async def stop(self):
+    """ Close the connection to the pump array. """
+    await self.halt()
+    assert self.modbus is not None, "Modbus connection not established"
+    self.modbus.close()
+    assert self.keep_alive_thread is not None, "Keep alive thread not established"
+    self.keep_alive_thread.join()
```

Comment:
I think we shouldn't assert that the modbus or thread exist before stopping. We just want to close them when they exist, but don't worry about the state where they are None (it could have failed to start), so I think we should say

```py
if self.modbus is not None:
  self.modbus.close()
if self.keep_alive_thread is not None:
  self.keep_alive_thread.join()
```

Note that the keep_alive_thread will not join, because it never quits. Instead, I would suggest adding a `_stopping_thread` (or similar) attribute, which you set to `True` right before the join, and rewriting `while True` to `while not _stopping_thread`. You can set it to `False` after the join succeeds.

Reply:
okay, I'll make these changes. I think I only added these asserts to make mypy happy.